tcp-transport/src/lib.rs

```diff
@@ -102,49 +124,85 @@ impl<Item, SinkItem> Transport<TcpStream, Item, SinkItem> {
 }
 
 /// Returns a new JSON transport that reads from and writes to `io`.
-pub fn new<Item, SinkItem>(io: TcpStream) -> Transport<TcpStream, Item, SinkItem>
+pub fn new<Item, ItemDecoder, SinkItem, SinkItemEncoder>(
+    io: TcpStream,
+    decoder: ItemDecoder,
+    encoder: SinkItemEncoder,
+) -> Transport<TcpStream, Item, ItemDecoder, SinkItem, SinkItemEncoder>
 where
     Item: for<'de> Deserialize<'de>,
+    ItemDecoder: Deserializer<Item>,
     SinkItem: Serialize,
+    SinkItemEncoder: Serializer<SinkItem>,
 {
-    Transport::from(io)
+    Transport::from((io, decoder, encoder))
 }
 
-impl<S: AsyncWrite + AsyncRead, Item: serde::de::DeserializeOwned, SinkItem: Serialize> From<S>
-    for Transport<S, Item, SinkItem>
+impl<S, Item, ItemDecoder, SinkItem, SinkItemEncoder> From<(S, ItemDecoder, SinkItemEncoder)>
```

Comment:
Just an FYI, this is the impl that allows users to provide non-TCP underlying transports.